librad/src/git/types/remote.rs

```diff
@@ -0,0 +1,164 @@
+}
+
+impl<Url> Remote<Url> {
+    /// Create a `"rad"` remote with a single fetch spec.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::marker::PhantomData;
+    /// use librad::{
+    ///     git::{
+    ///         local::url::LocalUrl,
+    ///         types::{remote::Remote, FlatRef, Force, NamespacedRef},
+    ///     },
+    ///     hash::Hash,
+    ///     uri::{Path, Protocol, RadUrn},
+    /// };
+    ///
+    /// let id = Hash::hash(b"geez");
+    ///
+    /// // The RadUrn pointing some project
+    /// let urn = RadUrn::new(
+    ///     id.clone(),
+    ///     Protocol::Git,
+    ///     Path::parse("").unwrap(),
+    /// );
+    ///
+    /// // The working copy heads, i.e. `refs/heads/*`.
+    /// let working_copy_heads: FlatRef<String, _> = FlatRef::heads(PhantomData, None);
+    ///
+    /// // The monorepo heads, i.e. `refs/namespaces/<id>/refs/heads/*`.
+    /// let monorepo_heads = NamespacedRef::heads(id, None);
+    ///
+    /// // Setup the fetch and push refspecs.
+    /// let fetch = working_copy_heads
+    ///         .clone()
+    ///         .refspec(monorepo_heads.clone(), Force::True)
+    ///         .into_dyn();
+    /// let push = monorepo_heads.refspec(working_copy_heads, Force::False).into_dyn();
+    ///
+    /// // We point the remote to `LocalUrl` which will be of the form `rad://<id>.git`.
+    /// let url: LocalUrl = urn.into();
+    ///
+    /// // Setup the `Remote`.
+    /// let mut remote = Remote::rad_remote(url, fetch);
+    /// remote.add_pushes(vec![push].into_iter());
+    /// ```
+    pub fn rad_remote(url: Url, fetch_spec: Box<dyn AsRefspec>) -> Self {
+        Self {
+            url,
+            name: "rad".to_string(),
+            fetch_spec,
+            push_specs: vec![],
+        }
+    }
+
+    /// Add a series of push specs to the remote.
+    pub fn add_pushes<I>(&mut self, specs: I)
+    where
+        I: Iterator<Item = Box<dyn AsRefspec>>,
+    {
+        for spec in specs {
+            self.push_specs.push(spec)
+        }
+    }
+
+    /// Create the [`git2::Remote`] and add the specs.
+    pub fn create<'a>(&self, repo: &'a git2::Repository) -> Result<git2::Remote<'a>, git2::Error>
+    where
+        Url: ToString,
+    {
+        let _ = repo.remote_with_fetch(
+            &self.name,
+            &self.url.to_string(),
+            &self.fetch_spec.as_refspec(),
+        )?;
+
+        for spec in self.push_specs.iter() {
+            repo.remote_add_push(&self.name, &spec.as_refspec())?;
+        }
+
+        // To ensure that the push spec is persisted we need to call `find_remote` here.
+        // Otherwise, `remote_add_push` doesn't affect the "loaded remotes".
+        repo.find_remote("rad")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{io, marker::PhantomData};
+
+    use super::*;
+    use crate::{git::types::*, hash::Hash};
+    use librad_test::tempdir::WithTmpDir;
+
+    #[test]
+    fn can_create_remote() {
```

Comment:
I noticed that we prefix tests with `test_`, which is missing in this case.

Reply:
Emmmm I'm not sure that's necessarily true. I think there are a mix and match. Since they're under `test` modules anyway the `test_` prefix is redundant and I didn't give that enough thought at the time :sweat_smile:

Reply:
I might have happen to check a file where all tests were prefixed with `test_` :P but agree about the redundancy, just wanted to point out a possible inconsistency.